include/dlaf/communication/communicator_pipeline.h

```diff
@@ -0,0 +1,141 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+/// @file
+
+#pragma once
+
+#include <dlaf/common/index2d.h>
+#include <dlaf/common/pipeline.h>
+#include <dlaf/common/round_robin.h>
+#include <dlaf/communication/communicator.h>
+
+namespace dlaf::comm {
+namespace internal {
+static constexpr const dlaf::common::Ordering FULL_COMMUNICATOR_ORDER{
+    dlaf::common::Ordering::RowMajor};
+}
+
+// TODO: Move this to separate header?
+/// TAG for strong-typing basic_coords.
+struct TAG_MPI;
+
+/// 2D index strong-typed for MPI.
+using Index2D = common::Index2D<IndexT_MPI, TAG_MPI>;
+/// 2D size strong-typed for MPI.
+using Size2D = common::Size2D<IndexT_MPI, TAG_MPI>;
+
+namespace internal {
+using CommunicatorPipeline = dlaf::common::Pipeline<Communicator>;
+}
+
+using CommunicatorPipelineReadOnlyWrapper =
+    typename internal::CommunicatorPipeline::ReadOnlyWrapper;
+using CommunicatorPipelineReadWriteWrapper =
+    typename internal::CommunicatorPipeline::ReadWriteWrapper;
+using CommunicatorPipelineReadOnlySender =
+    typename internal::CommunicatorPipeline::ReadOnlySender;
+using CommunicatorPipelineReadWriteSender =
+    typename internal::CommunicatorPipeline::ReadWriteSender;
+
+template <TODOCoord Coord> class CommunicatorPipeline {
+  using PipelineType = dlaf::common::Pipeline<Communicator>;
+
+public:
+  using ReadOnlyWrapper = typename PipelineType::ReadOnlyWrapper;
+  using ReadWriteWrapper = typename PipelineType::ReadWriteWrapper;
+  using ReadOnlySender = typename PipelineType::ReadOnlySender;
+  using ReadWriteSender = typename PipelineType::ReadWriteSender;
+
+  static constexpr TODOCoord coord = Coord;
+
+  /// Create a CommunicatorPipeline by moving in the resource (it takes the
+  /// ownership).
+  explicit CommunicatorPipeline(Communicator comm, Index2D rank = {0, 0},
+                                Size2D size = {0, 0})
+      : pipeline_(std::move(comm)), rank_(std::move(rank)),
+        size_(std::move(size)) {}
+  CommunicatorPipeline(CommunicatorPipeline &&other) = default;
+  CommunicatorPipeline &operator=(CommunicatorPipeline &&other) = default;
+  CommunicatorPipeline(const CommunicatorPipeline &) = delete;
+  CommunicatorPipeline &operator=(const CommunicatorPipeline &) = delete;
+
+  /// Return the rank of the current process in the CommunicatorPipeline.
+  ///
+  /// @return the 2D coordinate representing the position in the grid that this
+  /// pipeline belongs to
+  Index2D rank() const noexcept { return rank_; }
+
+  /// Return the size of the grid.
+  Size2D size() const noexcept { return size_; }
```

Comment:
Should these be independent or dependent on `TODOCoord`? Should there be two sets of functions: one for (potentially) 1D index/size (only `Row` and `Col`) and one for the full grid?